More Tests for WebRTC Data Channels

lgrahl opened the pull request:
### Description

This PR updates all data channel related tests to reflect the current spec (at the time of writing, that was [revision 1cc5bfc3ff18741033d804c4a71f7891242fb5b3](https://rawgit.com/w3c/webrtc-pc/1cc5bfc3ff18741033d804c4a71f7891242fb5b3/webrtc.html)).

### Updated/Fixed Tests

* RTCDataChannel-id
  - *DTLS client uses odd data channel IDs*
  - *DTLS server uses even data channel IDs*
* RTCDataChannel-send-receive (renamed from RTCDataChannel-send)
  - (\<type> is of *In-band negotiated channel*, *Negotiated channel* or *Asymmetric negotiated channel*)
  - *\<type>: Calling send() when not open should throw InvalidStateError* (renamed from *Calling send() when data channel is in connecting state should throw InvalidStateError*)
  - *\<type>: Calling send() with null should throw TypeError*
  - *\<type>: Calling send() with undefined should throw TypeError*
  - *\<type>: Should be able to send (local) simple string and receive (remote) as string* (based on *Data channel should be able to send simple string and receive as string*)
  - *\<type>: Should be able to send (remote) simple string and receive (local) as string* (based on *Data channel should be able to send simple string and receive as string*)
  - *\<type>: Should be able to send unicode string and receive as unicode string* (renamed from *Data channel should be able to send unicode string and receive as unicode string*)
  - *\<type>: Should ignore binaryType and always receive string message as string* (renamed from *Data channel should ignore binaryType and always receive string message as string*)
  - *\<type>: Should be able to send Uint8Array message and receive as ArrayBuffer* (renamed from *Data channel should be able to send Uint8Array message and receive as ArrayBuffer*)
  - *\<type>: Should be able to send ArrayBuffer message and receive as ArrayBuffer* (renamed from *Data channel should be able to send ArrayBuffer message and receive as ArrayBuffer*)
  - *\<type>: Should be able to send Blob message and receive as ArrayBuffer* (renamed from *Data channel should be able to send Blob message and receive as ArrayBuffer*)
  - *\<type>: Should be able to send ArrayBuffer message and receive as Blob* (renamed from *Data channel should be able to send ArrayBuffer message and receive as Blob*)
  - *\<type>: Should be able to transmit an empty string* (moved from and based on datachannel-emptystring *Can send empty strings across a WebRTC data channel.*)
  - *\<type>: Should receive messages as Blob by default* (renamed from *Data channel binaryType should receive message as Blob by default*)
  - *\<type>: Sending multiple messages with different types should succeed* (based on *Sending multiple messages with different types should succeed and be received*)
  - *\<type>: Receiving multiple messages with different types should succeed* (based on *Sending multiple messages with different types should succeed and be received*)
* RTCPeerConnection-createDataChannel
  - *createDataChannel attribute default values*
  - *createDataChannel with provided parameters should initialize attributes to provided values*
  - *createDataChannel with both maxPacketLifeTime and maxRetransmits should throw TypeError* (renamed from *createDataChannel with both maxPacketLifeTime and maxRetransmits should throw SyntaxError*)
  - *createDataChannel with too long label should throw TypeError* (renamed from *createDataChannel with negotiated false and long label should throw TypeError*)
  - *createDataChannel with too long protocol should throw TypeError* (renamed from *createDataChannel with negotiated false and long protocol should throw TypeError*)
  - *Channels created (after setRemoteDescription) should have id assigned* (renamed from *Channels created after SCTP transport is established should have id assigned*)
* RTCPeerConnection-ondatachannel
  - *Data channel event should fire when new data channel is announced to the remote peer* (renamed from *datachannel event should fire when new data channel is announced to the remote peer*)
  - *In-band negotiated channel created on remote peer should match the same configuration as local peer* (renamed from *Data channel created on remote peer should match the same configuration as local peer*)
* RTCSctpTransport-constructor
  - *setRemoteDescription() with answer containing data media should initialize pc.sctp*
  - *setLocalDescription() with answer containing data media should initialize pc.sctp*
* RTCSctpTransport-maxMessageSize
  - *Remote offer SDP missing max-message-size attribute*
  - *max-message-size with a (non-zero) value provided by the remote peer*
  - *Renegotiate max-message-size with various values provided by the remote peer* (based on *Renegotiate max-message-size with a (non-zero) value provided by the remote peer*)
  - *max-message-size with a (non-zero) value larger than canSendSize provided by the remote peer*

### New Tests

* RTCDataChannel-binaryType
  - *Setting binaryType to 'blob' should succeed*
  - *Setting binaryType to 'arraybuffer' should succeed*
  - *Setting invalid binaryType 'jellyfish' should throw SyntaxError*
  - *Setting invalid binaryType 'arraybuffer ' should throw SyntaxError*
* RTCDataChannel-bufferedAmount
  - *bufferedAmount initial value should be 0 for both peers*
  - *bufferedAmount should not decrease immediately after initiating closure*
  - *bufferedAmount should not decrease after closing the peer connection*
* RTCDataChannel-bufferedAmountLowThreshold
  - (\<type> is of `DOMString` or `Uint8Array`)
  - *\<type> of 4 bytes: bufferedamountlow event should fire once bufferedAmount <= 0 (default)*
  - *\<type> of 4 bytes: bufferedamountlow event should fire once bufferedAmount <= 2 (custom)*
  - *\<type> of 4 bytes: bufferedamountlow event should fire once bufferedAmount <= 4 (custom)*
  - *\<type> of 4 bytes: bufferedamountlow event should fire once bufferedAmount <= 8 (custom)*
  - *\<type> of 4 bytes x 10: bufferedamountlow event should fire once bufferedAmount <= 2 (custom)*
  - *\<type> of 4 bytes x 10: bufferedamountlow event should fire once bufferedAmount <= 4 (custom)*
  - *\<type> of 4 bytes x 10: bufferedamountlow event should fire once bufferedAmount <= 8 (custom)*
  - *\<type> of 4 bytes: bufferedamountlow event should not fire on other channels*
* RTCDataChannel-close
  - *In-band data channel closing before connection establishment*
  - *In-band data channel closing after connection establishment*
  - *In-band data channels x10 closing on local side should be synchronized across peers*
  - *In-band data channels x10 closing on remote side should be synchronized across peers*
  - *Negotiated data channels x10 closing should be synchronized across peers*
  - *Closing peer connection locally should close all 10 channels for the remote peer*
* RTCDataChannel-dcep
  - (\<value> is of `undefined`, `null`, `0`, `65535` or `4294967295`)
  - *Ordered and reliable channel should be created via DCEP*
  - *Unordered and reliable channel should be created via DCEP*
  - *Ordered and unreliable (maxRetransmits=\<value>) channel should be created via DCEP*
  - *Unordered and unreliable (maxRetransmits=\<value>) channel should be created via DCEP*
  - *Ordered and unreliable (maxPacketLifeTime=\<value>) channel should be created via DCEP*
  - *Unordered and unreliable (maxPacketLifeTime=\<value>) channel should be created via DCEP*
  - *Channel with priority set to high should be created via DCEP*
  - *Channel IDs should be synchronized when created via DCEP*
  - *Zero length label and protocol option should be transmitted via DCEP*
  - *Maximum length label and protocol option should be transmitted via DCEP*
  - *Maximum length label and protocol option (3 byte unicode) should be transmitted via DCEP*
* RTCDataChannel-id
  - *In-band negotiation with a specific ID should not be allowed*
  - *Odd/even role should not be violated when mixing with negotiated channels*
  - *Creating a channel after exhausting the maximum number of channels should throw OperationError (after connection establishment)*
  - *ID reuse should not be possible before the former channel with the same ID closed*
  - *ID reuse should be possible once the former channel with the same ID closed*
  - *Channel IDs should be reassigned after exhaustion in case a channel has been closed*
  - *Exhausting channels with one peer should not violate the odd/even rule*
  - *Channel ID exhaustion handling (before and after connection establishment)*
* RTCDataChannel-onopen
  - *In-band channel: Open event should be fired (local) when the data channel opens*
  - *In-band channel: Open event should be fired (remote) when the data channel opens*
  - *Negotiated channel: Open event should be fired when the channels open*
* RTCDataChannel-send-receive
  - (\<type> is of *In-band negotiated channel*, *Negotiated channel* or *Asymmetric negotiated channel*)
  - (\<size> is of `1024`, `16384`, `65536`, `131072`, `262144`, `524288`, `1048576`, `16777216` or `33554432`)
  - *\<type>: Calling send() with 42 should throw TypeError*
  - *\<type>: Should be able to send Int8Array message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Uint8ClampedArray message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Int16Array message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Uint16Array message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Int32Array message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Uint32Array message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Float32Array message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Float64Array message and receive as ArrayBuffer*
  - *\<type>: Should be able to send DataView message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Uint8Array (with offset) message and receive as ArrayBuffer*
  - *\<type>: Should be able to send DataView (with offset) message and receive as ArrayBuffer*
  - *\<type>: Should be able to send Blob (with offset) message and receive as ArrayBuffer*
  - *\<type>: Should be able to transmit an empty Uint8Array*
  - *\<type>: Should be able to transmit an empty Blob*
  - *\<type>: Sending and receiving \<size> bytes should succeed or raise TypeError*
  - *\<type>: Sending and receiving \<size> +1 bytes should raise TypeError*
  - *\<type>: Sending and receiving 16 KiB x64 should succeed*
  - *\<type>: Sending and receiving 16 KiB x256 on both peer simultaneously should succeed*
  - *\<type>: Sending and receiving 2048 messages should succeed and keep order*
  - *\<type>: Closing a channel (local) with pending data should transfer that data before becoming closed*
  - *\<type>: Sending after the channel has been closed (local) should raise InvalidStateError*
  - *\<type>: Sending after the channel has been closed (remote) should raise InvalidStateError*
  - *\<type>: Sending after the peer connection has been closed (local) should raise InvalidStateError*
  - *\<type>: Sending after the peer connection has been closed (remote) should raise InvalidStateError*
* RTCDataChannel-send-receive-stress
  - *Stress-test with multiple channels sending and receiving using various data sources*
* RTCPeerConnection-close
  - *Closing the peer connection should close all channels (before connection establishment)*
  - *Closing the local peer connection should close all channels (after connection establishment)*
  - *Closing the remote peer connection should close all channels (after connection establishment)*
  - *Closing the peer connection should set the SCTP transport to closed*
* RTCPeerConnection-createDataChannel
  - *createDataChannel with both maxPacketLifeTime and maxRetransmits null should succeed*
  - *createDataChannel with too long label (2 byte unicode) should throw TypeError*
  - *createDataChannel with same label used twice should not throw*
  - *createDataChannel with too long protocol (2 byte unicode) should throw TypeError*
  - *createDataChannel with maximum length label and protocol should succeed*
  - *createDataChannel with negotiated false should succeed*
  - *createDataChannel with negotiated false and id 42 should ignore the id*
  - *createDataChannel with negotiated true and id not defined should throw TypeError*
  - *createDataChannel with negotiated true and id null should throw TypeError*
  - *Reusing a data channel id that is in use should throw OperationError*
  - *Reusing a data channel id that is in use (after setRemoteDescription) should throw OperationError*
  - *Reusing a data channel id that is in use (after setRemoteDescription, negotiated via DCEP) should throw OperationError*
  - *New data channel should be in the connecting state after creation (after connection establishment)*
* RTCPeerConnection-ondatachannel
  - *Should be able to send data in a datachannel event handler*
  - *Open event should not be raised when closing the channel in the datachannel event*
  - *Open event should be raised when closing the channel in the datachannel event after enqueuing a task*
  - *Open event should not be raised when sending and immediately closing the channel in the datachannel event*
  - *In-band negotiated channel created on remote peer should match the same (default) configuration as local peer*
* RTCSctpTransport-constructor
  - *setRemoteDescription() with answer not containing data media should not initialize pc.sctp*
  - *setLocalDescription() with answer not containing data media should not initialize pc.sct*
* RTCSctpTransport-onstatechange
  - *RTCSctpTransport.onstatechange should fire when the state changes*
* historical
  - *RTCDataChannel member maxRetransmitTime should not exist*

### Comments

Only few helper functions have been changed but everything is either backwards compatible or the affected tests have been updated. FWIW, I have carefully tested new test cases (if that was possible) on Firefox and Chrome by temporarily commenting out asserts and in some cases even by throwing in adapter (`RTCSctpTransport.maxMessageSize` for example).

Towards WebRTC implementations: A bunch of new issues are revealed by these tests. I'm at a tight time schedule, so I can't file them for you at the moment. This listing here should help you in finding new or updated test cases, so you can track and file new issues in your bug trackers. /cc @nils-ohlmeier @jan-ivar. Please also forward this to whoever is responsible for Google's, Safari's and Edge's (:joy_cat:) data channel implementation.

And I apologise for creating this rather large PR.

lgrahl committed:
Update RTCPeerConnection-createDataChannel tests

Update to current spec version
Fix missing import to RTCPeerConnection-helper.js
Test .bufferedAmount and .bufferedAmountLowThreshold
Fix invalid use of 'label' in RTCDataChannelInit
Add negotiated true and id null test
Add reusing an id that is in use tests

lgrahl committed:
Update RTCPeerConnection-id to current spec version (only the
description has been updated)

lgrahl committed:
Add RTCPeerConnection-dcep tests which test whether data channel
options are synchronized correctly across peers.

Add optional 'channelLabel' and 'channelOptions' to `createDataChannelPair`

lgrahl committed:
Update RTCSctpTransport-maxMessageSize tests

Once canSendSize has been determined, require it to be available in further tests that depend on it
Fix canSendSize checks (explicitly check for the 0 case)
Fix remoteMMS > canSendSize test for the 0 case
Update descriptions to make them more precise

lgrahl committed:
Add exhausting ID tests to RTCSctpTransport-id tests

Add a test that tries to create the maximum number of data channels
until no ids are free anymore after the connection has been established.
Add a test that checks if channel IDs are being reassigned after
exhaustion in case a channel has been closed.
Add a test that tries to create the theoretical maximum number of data
channels and ensure that violation of the odd/even rule does not occur.
Add a test that tries to create the theoretical maximum number of data
channels before the connection has been established (and check if some
of them are properly teared down).

lgrahl committed:
Update the renegotiation test in RTCSctpTransport-maxMessageSize

lgrahl committed:
Fix RTCSctpTransport-id tests to always check for reused IDs
Add test for violation of odd/even rule with mixed channels (negotiated & DCEP-negotiated)

lgrahl committed:
Add tests for data channel closing procedure when the peer connection is being closed

lgrahl committed:
Fix do not attempt to keep data channels referenced since they should have a strong reference to the RTCPeerConnection instance anyway

lgrahl committed:
Fix title of DCEP tests
Move RTCPeerConnection.close tests into their own file

lgrahl committed:
Update RTCSctpTransport-constructor tests

Update to current spec version
Add assert for 'state'
Fix the assert for 'maxMessageSize'
A bit of cleanup

lgrahl committed:
Add RTCSctpTransport-onstatechange test which ensures that the event fires with the correct state

Add an `IceCandidateQueue` helper class which allows you hold back ICE candidates conveniently until `exchangeIceCandidates` is called

lgrahl committed:
Add test to ensure the SCTP transport attribute remains null in case no data channels are being created

lgrahl committed:
Add test to ensure the SCTP transport attribute remains null in case no SCTP transport is being negotiated

lgrahl committed:
Add test for SCTP transport closed state when the peer connection is being closed

lgrahl committed:
Add RTCDataChannel-onopen test

lgrahl committed:
Update RTCPeerConnection-ondatachannel to the current spec (descriptions and comments only)

lgrahl committed:
Add RTCDataChannel-close which includes tests regarding the closing procedure and the onclose event
Add some timeout and fix missing unreachable checks for RTCDataChannel-onopen

lgrahl committed:
Add untestable note to RTCDataChannel-close for data channel ids outside of the range of the negotiated amount of SCTP streams

lgrahl committed:
Add test for RTCDataChannel label used twice

lgrahl committed:
Update the description of RTCDataChannel-bufferedAmount

lgrahl committed:
Add RTCDataChannel-bufferedAmountLowThreshold which tests the bufferedamountlow event and the threshold value itself

lgrahl committed:
Fix a bunch of invalid 'done' timeouts I've added
Fix missing assert_unreached in RTCDataChannel-close
Fix invalid creation of a data channel in RTCDataChannel-close

lgrahl committed:
Update RTCDataChannel-send to the current spec

Add send test for DataView
Add test for invalid binary type
Add test for changing the binary type on-the-fly
Add tests for sending and receiving n bytes
Add test for closing a channel with outstanding data

lgrahl committed:
Refactor and move empty string/buffer/blob tests into RTCDataChannel-send
Remove datachannel-emptystring

lgrahl committed:
Add test to RTCDataChannel-send which patches the remote maximum message size and expects a TypeError in case the message is larger than that

lgrahl committed:
Rename RTCDataChannel-send to RTCDataChannel-send-receive

lgrahl committed:
Add spec revision to RTCDataChannelEvent-constructor
Minor refactoring

lgrahl committed:
Add ID reuse test after close to RTCDataChannel-id
Split sending outstanding messages before a channel may become closed tests into two (close on the local/remote side) in RTCDataChannel-send-receive
Add DCEP-channel ID reuse test to RTCPeerConnection-createDataChannel

lgrahl committed:
Use different message sizes for tests in RTCDataChannel-bufferedAmountLowThreshold
Add a test which enqueues 10 messages

lgrahl committed:
Add timeout to tests in RTCDataChannel-bufferedAmountLowThreshold

lgrahl committed:
Update RTCDataChannel-close tests to close multiple channels
Add a test to RTCDataChannel-close which tests for the onclose event when closing the peer connection
Add possibility to create multiple pairs when using the createDataChannelPair helper function
Update all tests that use that helper function

lgrahl committed:
Update RTCPeerConnection-ondatachannel, readyState should be 'connecting'

lgrahl committed:
Add test for checking the ready state of a newly created data channel after the connection has been established

lgrahl committed:
Fix missing id when using negotiated: true for data channel long label/protocol tests

lgrahl committed:
Add test for data channel ID set to a number but negotiated set to false

lgrahl committed:
Add test for maximum length label and protocol using 3 byte unicode when creating a data channel via DCEP

lgrahl committed:
Add tests for too long protocol and too long label by using unicode characters when creating a data channel

lgrahl committed:
Add more reliability parameter tests in RTCDataChannel-dcep

lgrahl committed:
Add non-zero test case for creating a data channel with maxPacketLifeTime and maxRetransmits set

lgrahl committed:
Add test case for createDataChannel with maxPacketLifeTime and maxRetransmits being null

lgrahl committed:
Fix the ondatachannel event test (don't set the ID, fix ID comparison, add negotiated flag check)

lgrahl committed:
Add test for default data channel configuration sync for the 'ondatachannel' event

lgrahl committed:
Add test for in-band negotiation of a data channel with a specific ID (should not be allowed)

lgrahl committed:
Run all send/receive tests for negotiated channels as well

lgrahl committed:
Run all send/receive tests for negotiated asymmetric channels as well

lgrahl committed:
Add RTCDataChannel.maxRetransmitTime member to historical

lgrahl committed:
Split "closing the peer connection should close all channels" test into local and remote

lgrahl committed:
Update exhausting channel IDs with one peer test to create 66535 data channel

lgrahl committed:
Add todo note for https://github.com/w3c/webrtc-pc/issues/1829

lgrahl committed:
Add test for channel ID reuse after a former data channel with the same ID closed

lgrahl committed:
Split onopen test case into local and remote (for in-band negotiated channels)
Add onopen event test case for negotiated channel

lgrahl committed:
Add test for sending a message on the remote side (where the channel has been opened from the local side)

lgrahl committed:
Add test for sending 16 KiB x64 and check if it is being received

lgrahl committed:
Add tests for sending after a channel (or a peer connection) has been closed

lgrahl committed:
Add a test that ensures the initial .bufferedAmount value is zero for both peers

lgrahl committed:
Add test for data channel buffered amount after closing a channel or a peer connection
Minor refactoring and cleanup of RTCDataChannel-bufferedAmount

lgrahl committed:
Add test cases for sending all TypedArray types
Add test cases for sending all types with offsets
Fix tests for sending after closure of a channel (or a peer connection)
Rename `assert_equals_array_buffer` and refactor to `assert_equals_typed_array` which accepts all kind of typed arrays and compares them using a DataView

lgrahl committed:
Split send after close (channel and peer connection) into local and remote

lgrahl committed:
Add data channel test case where both peers send simultaneously

lgrahl committed:
Add data channel test case for ensuring an ordered data channel is ordered

lgrahl committed:
Enhance descriptions of some test cases
Fix typo

lgrahl committed:
Fix SCTP transport constructor tests

lgrahl committed:
Add a short stress test for data channel sending/receiving
Remove some console logging the linter complains about

lgrahl committed:
Add license since some of the data channel tests are based on or inspired by previous work

lgrahl committed:
Add data channel send tests for `null` and `undefined`

lgrahl committed:
Update all data channel tests to use async/await
Make all data channel tests based on `promise_test` (to prevent resource exhaustion)
Some cleanup

lgrahl committed:
Update `datachannel` event tests for https://github.com/w3c/webrtc-pc/pull/1851

lgrahl committed:
Add a test based on discussion in https://github.com/w3c/webrtc-pc/pull/1851 with @jan-ivar

lgrahl committed:
Remove invalid *negotiated true and no id* test case
Add explicit *negotiated false* and *negotiated true and id not defined* test cases

lgrahl committed:
Add RTCDataChannel-binaryType test file
Move an existing test case from RTCDataChannel-send-receive

lgrahl committed:
Remove a test case that is invalid now since https://github.com/w3c/webrtc-pc/pull/1849 has been merged

lgrahl committed:
Add the number 42 to the list of invalid types when sending a data channel message

lgrahl committed:
Fix missing local timeout

lgrahl committed:
Add an ugly workaround for createDataChannelPair to make it work in Safari

lgrahl committed:
Increase timeout for channel exhaustion tests

lgrahl committed:
Fix assert_unreached description

lgrahl committed:
Fix missing attribute check and a typo in RTCPeerConnection-ondatachannel

lgrahl committed:
Fix a typo

lgrahl committed:
Fix test description

lgrahl committed:
Make sure to skip at least a task in the `datachannel` then `open` event test

lgrahl committed:
Fix bufferedamountlow event test

lgrahl committed:
Update Resolver class to extend a Promise
Use Resolver class in data channel tests and update other usages

lgrahl committed:
Minor fixes after rebasing

lgrahl committed:
Make the linter happy

lgrahl commented:
Harald, can I ping you for another review? The changes to testharness have been removed and the tests have been rebased against master.

/cc @alvestrand

foolip commented:
@lgrahl thank you for contributions! Unfortunately, I have to ask that you not add a license file to the repo beyond the existing https://github.com/web-platform-tests/wpt/blob/master/LICENSE.md. There's an ongoing discussion about changing it in https://github.com/web-platform-tests/wpt/issues/11009 but for now the license is what it is.

If you don't yourself have the rights to all of this code, then please work with the original authors to approve contribution under the license of this repo.

lgrahl commented:
@foolip All authors approve to publish under the W3C license but I would like to keep the attribution to these three guys. Can I change the license file to an attribution file?

foolip commented:
If you'd like to acknowledge the authors, I'd suggest listing them in the tests, like this:
https://github.com/web-platform-tests/wpt/blob/7c50c216081d6ea3c9afe553ee7b64534020a1b2/css/css-transforms/transform-compound-001.html

lgrahl commented:
I would like to do that but associating the authors with the individual test cases (and files) is near impossible by now (I can explain why if you want me to). Adding them to all data channel files is also a bit weird but I guess I wouldn't mind. What do you think?

agouaillard-cosmo commented:
The license and CLA does not allow you to touch license and copyright, so In my humble opinion, foolip is right, and you should put their name in every single file you think deserve the mention.

foolip commented:
Oh, all right, well in that case perhaps you can put something in a README.md. https://github.com/web-platform-tests/wpt/commit/dd299f45326e12acf453a14db5b582eba6d5db63 is another README that explains the origins of something, as an example.

agouaillard-cosmo commented:
if you put something in the readme, you will need to put the list of file those names are credited in the readme. The readme then becomes dependent on those files, and you need to know to update it when you modify the files. It's brittle in my opinion. Better to keep the credits in corresponding files.

foolip commented:
Getting attribution exactly right and having it stay right over time is a pain, so I'd just say "some tests". Sort of like some of the attributions after the long list of names in https://html.spec.whatwg.org/multipage/acknowledgements.html, which aren't very precise about exactly what came from whom.

lgrahl committed:
Add readme file to WebRTC test suite and provide attribution

Remove license file

lgrahl commented:
I have removed the license file and added a readme (with attribution).

I must say I agree with @foolip here as the attribution cannot be tied to specific files, so we just shouldn't try. Just to explain why this happened: First of all, I have updated all existing data channel tests to be up to date. Then, to maximise diversity and not spoil myself, I have written my own tests based on the specification and on existing examples I've written a while ago. After that, I have added missing tests from the [data channel conformance test suite][nplab-webrtc-dc-playground]. And I don't remember which ones that were because I didn't think it would be important. Anyhow, because of this, saying that "some data channel tests are based on ..." seems fair to me.

[nplab-webrtc-dc-playground]: https://github.com/nplab/WebRTC-Data-Channel-Playground/tree/master/conformance-tests

foolip commented on webrtc/README.md:
```diff
@@ -0,0 +1,12 @@
+
+This directory contains the WebRTC test suite.
+
+## Attribution
```
This looks great! To avoid the possible impression that this is fulfilling a license requirement for attribution (such as with CC-by) may I suggest "Acknowledgments" instead?

foolip commented on webrtc/README.md:
```diff
@@ -0,0 +1,12 @@
+Some data channel tests are based on the [data channel conformance test
+suite][nplab-webrtc-dc-playground] of the Network Programming Lab of the Münster
+University of Applied Sciences. We would like to thank Peter Titz, Felix Weinrank and Timo
+Völker for allowing us to make their test cases available under the W3C license.
```
Note that per https://github.com/web-platform-tests/wpt/issues/11009 the license may change, as allowed for by the terms in CONTRIBUTING.md.

To not have to update this text later, may I suggest not mentioning the license but instead sometime like "agreeing to contribute their test cases to this repository"?

youennf commented on webrtc/RTCDataChannel-bufferedAmount.html:
```diff
@@ -5,165 +5,239 @@
+  assert_equals(dc1.bufferedAmount, 0, 'Expect bufferedAmount to be 0');
+  assert_equals(dc2.bufferedAmount, 0, 'Expect bufferedAmount to be 0');
+}, 'bufferedAmount initial value should be 0 for both peers', {
+  timeout: 5000
```
Even for async tests, I am not sure this is worth the effort to add specific timeouts.
In some configurations, 5 seconds will be what is expected to run the tests of the whole file for instance.

If we really want to do something there, maybe createDataChannelPair should actually reject after some timeout?
That way, if we have several await in a promise test, we will know which one is timing out.

lgrahl replied:
If we don't specify the local timeout, the test case will block running all other test cases. The alternative is to carefully wrap all possible awaitables with timeouts but I don't want to go that way for several reasons:

- it's quite a lot of boilerplate,
- makes the test cases brittle since people (including me) will tend to forget an awaitable here and there, and
- not every awaitable that could in theory time out is likely to time out.

Adding failure context to some awaitables by using timeouts is a good idea, though (and I have done that in vital places). Rejecting `createDataChannelPair` after some timeout could be done but since that isn't part of the test case definition, the question is when the timeout should fire...

lgrahl replied:
What about the tests that share 99% of their code?

youennf replied:
Sure, if some tests share 99%, it is ok to group them (and incur the timeout risk).

The alternative is of course to have some helper functions in some JS files.
That way, you keep the meat of each subtest in each promise_test function.
And you can then split tests as desired.

In general, the more tests on a given file, the more flaky.
And the more difficult to debug as well...

youennf commented on webrtc/RTCPeerConnection-helper.js:
```diff
@@ -191,93 +191,242 @@ function test_never_resolve(testFunc, testName) {
   pc1=new RTCPeerConnection(),
-  pc2=new RTCPeerConnection())
+  pc2=new RTCPeerConnection(),
+  options={})
```
s/=/ = /

youennf reviewed (commented):
This looks great, thanks a lot!

I haven't looked at whether the tests are matching the spec or not.
Given the number of tests, we might want to concentrate on the tests that show differences/bugs between implementation and spec. If you already have a list, that would be great, otherwise it should be ok to repro it locally.

Also, I am not a fan of explicit handling of timeouts, I would tend to remove most of them.

alvestrand commented:
@lgrahl can we find some time when you'll be available for updating the tests? Having to wait multiple days after a review before seeing the responses increases the cognitive load on reviewers - if we can find a day to sit down and ping-pong on these tests, it should be possible to get the review done.
Also, Chrome tests have trouble - see bot run.

lgrahl commented:
@alvestrand @youennf Sorry for the delay. I'm not sure what the issue is on Travis since the Taskcluster seems to run fine. Looking at the log, I'm speculating that Travis takes too long to process the large changeset... but I might be wrong. In any way, I was not able to pin it down so far.

I'm available today for *ping-pong*. The upcoming weeks are packed.

lgrahl committed:
Update readme (attribution -> acknowledgements)

lgrahl committed:
Update readme (licensing stuff)

lgrahl committed:
Remove timeouts from .binaryType tests

lgrahl commented on webrtc/RTCDataChannel-binaryType.html:
```diff
@@ -0,0 +1,69 @@
+    dc.binaryType = binaryType;
+    assert_equals(dc.binaryType, binaryType, `dc.binaryType should be '${binaryType}'`);
+  }, `Setting binaryType to '${binaryType}' should succeed`, {
+    timeout: 3000
```
You're right, in this file the tests are not async so the timeouts can be removed. Done!

lgrahl commented on webrtc/RTCDataChannel-binaryType.html:
```diff
@@ -0,0 +1,69 @@
+
+// Disable global timeout
+// IMPORTANT: You need to add a timeout to *every test* of this file!
+setup({ explicit_timeout: true });
```
Removed (in this file).

lgrahl replied:
However, in general this exists to allow a per-test timeout instead of the inflexible global per-file timeout which can be either 10 seconds or 60 seconds. This is insufficient for a test file with many promise-based test cases that can only run sequentially. Therefore, these files provide a timeout for each test case and disable the global timeout.

lgrahl replied:
> If we fear that timing out subtests is an issue, we should simply split subtests in different files.

Don't forget that a `promise_test` blocks all subsequent tests. This means, if a `promise_test` does not have a local timeout and it does time out, it prevents all other tests from running. If we don't want local timeouts, then we have to move each test case into its own file.

youennf replied:
In practice, timing out subtests do happen.
Browser implementors tend to try fixing the corresponding issues sooner than regular failures.

lgrahl replied:
I very much like the thought of browser vendors fixing their implementations due to tests timing out but don't consider that approach a good thing. Anyhow, I have moved this discussion further down into the comments since I want to hear some more opinions on that.

lgrahl committed:
Make human linters happy

foolip commented:
Travis wasn't timing out, the limit is 50 minutes and the stability jobs took much less than that. Probably the tests actually are flaky.

lgrahl commented:
@foolip Do you have any hints? Frankly, I'm having a hard time to figure out what Travis is doing.

youennf commented:
I do not see the logs.
There seems to be some conflicts with the recent WebRTC patches we uploaded during today hackathon.

lgrahl commented:
There's an older build that failed (only cosmetic changes have been done in the meantime): https://travis-ci.org/web-platform-tests/wpt/builds/444438000

youennf commented:
For the record, the tests that seem to fail are:
1. "/webrtc/RTCDataChannel-bufferedAmountLowThreshold.html#DOMString of 4 bytes x 10: bufferedamountlow event should ...".
In WebKit, we had issues getting a consistent bufferedamountlow testing, we ended up writing a large enough message. Maybe this is a bug in implementations...
Isolating a flaky subtest from other subtests is good. Otherwise it might end up making other subtest results difficult to use.

2. "/webrtc/RTCDataChannel-close.html#Negotiated data channels x10 closing should be synchronized across peers"
Not sure why it fails. If nChannels is equal to 1, then nCloses will be 1 but will be compared to 0?
Should it be compared to nChannels? Ditto in some other places.

3. "/webrtc/RTCDataChannel-send-receive.html#In-band negotiated channel..."
Not sure why it fails either.
This test has both a step_timeout and an explicit timeout value, is there a particular reason for that.

The fact that there is a comment stating that a 10 seconds step_timeout is required worries me a bit.

I also spotted some very big timeouts at least in webrtc/RTCDataChannel-send-receive-stress.html
and webrtc/RTCDataChannel-send-receive.html (69 seconds, 420 seconds).
I would try either rewriting these tests or putting them aside in their own file.
These tests might otherwise be skipped in browsers CI, at least in slow configurations(think debug+asan with multiple tests run in parallel).

youennf commented:
For the record, WebKit default timeout for running a file is 30 seconds.
We can extend it to 150 seconds but we usually prefer not to have to do that.

lgrahl commented:
> 1. "/webrtc/RTCDataChannel-bufferedAmountLowThreshold.html#DOMString of 4 bytes x 10: bufferedamountlow event should ...".
> In WebKit, we had issues getting a consistent bufferedamountlow testing, we ended up writing a large enough message. Maybe this is a bug in implementations...

That bug is explained in my thesis but this being flaky is new to me. This should only become flaky if one reuses an existing SCTP association (`RTCPeerConnection` instance, or creating more than one data channel) in another test.

> [...] Isolating a flaky subtest from other subtests is good. Otherwise it might end up making other subtest results difficult to use.

I don't understand why it would make the results difficult to use. They don't depend on each other. If some test case breaks another, let me know because that would be a bug.

Splitting them however would duplicate a lot of code. I find the idea awkward to split test cases depending on the implementation status in browsers.

>     2. "/webrtc/RTCDataChannel-close.html#Negotiated data channels x10 closing should be synchronized across peers"
>        Not sure why it fails. If nChannels is equal to 1, then nCloses will be 1 but will be compared to 0?
>        Should it be compared to nChannels? Ditto in some other places.

Looks like a bug. I'll look into it.

>     3. "/webrtc/RTCDataChannel-send-receive.html#In-band negotiated channel..."
>        Not sure why it fails either.
>        This test has both a step_timeout and an explicit timeout value, is there a particular reason for that.

The only reason is being able to get the failure context.

> The fact that there is a comment stating that a 10 seconds step_timeout is required worries me a bit.

Mhh, but what's the alternative? Let it time out early to get Chrome people look into it?

> I also spotted some very big timeouts at least in webrtc/RTCDataChannel-send-receive-stress.html
> and webrtc/RTCDataChannel-send-receive.html (69 seconds, 420 seconds).
> I would try either rewriting these tests or putting them aside in their own file.

I'm not sure how to rewrite them... it just can take a while on a slow machine to send 32 MiB and receive it. But I absolutely think the test is necessary.

The stress test is already in its own file.

> These tests might otherwise be skipped in browsers CI, at least in slow configurations(think debug+asan with multiple tests run in parallel).

Is there a meta tag to mark it as a *long* test to be skipped by slow configurations?

lgrahl committed:
Fix RTCDataChannel-close `nCloses` vs. `nChannels` comparison

youennf commented:
> Is there a meta tag to mark it as a _long_ test to be skipped by slow configurations?

Yes: '<meta name="timeout" content="long">'
But stress tests will probably not work well in the way WebKit runs WPT tests in its CI.

youennf commented:
> > [...] Isolating a flaky subtest from other subtests is good. Otherwise it might end up making other subtest results difficult to use.
> 
> I don't understand why it would make the results difficult to use. They don't depend on each other. If some test case breaks another, let me know because that would be a bug.

In WebKit CI (Chrome might still be fairly similar), we generate a text file containing all the subtest results (usually a file containing one PASS/FAIL line per subtest). If one of the line is different, the whole test is marked as failed. If there are variations run after run in the subtest results, we mark the whole file as flaky, even though only one subtest is actually flaky. We do not have per subtest expectation.

youennf commented:
> > I also spotted some very big timeouts at least in webrtc/RTCDataChannel-send-receive-stress.html
> > and webrtc/RTCDataChannel-send-receive.html (69 seconds, 420 seconds).
> > I would try either rewriting these tests or putting them aside in their own file.
> 
> I'm not sure how to rewrite them... it just can take a while on a slow machine to send 32 MiB and receive it. But I absolutely think the test is necessary.

What exactly does it test? Is it the robustness of the implementation or the compliance of it?

webrtc/RTCDataChannel-send-receive-stress.html is just one test so I am not sure how feasible splitting is. I would go with just meta/long there.

lgrahl commented:
> > > [...] Isolating a flaky subtest from other subtests is good. Otherwise it might end up making other subtest results difficult to use.
> > 
> > 
> > I don't understand why it would make the results difficult to use. They don't depend on each other. If some test case breaks another, let me know because that would be a bug.
> 
> In WebKit CI (Chrome might still be fairly similar), we generate a text file containing all the subtest results (usually a file containing one PASS/FAIL line per subtest). If one of the line is different, the whole test is marked as failed. If there are variations run after run in the subtest results, we mark the whole file as flaky, even though only one subtest is actually flaky. We do not have per subtest expectation.

I see. No disrespect, but that seems like a design flaw of the *flaky* evaluation. Again, I'm really unhappy with that approach... it feels wrong to move test cases out of their test group because some implementation is incorrect. Moving test case definitions out of the actual test case to reuse it in both files doesn't feel any less wrong.

But I suppose this isn't going to change in just a few days, so let's go through each test case that fails for Chrome (since I can't see which ones fail for Firefox as the log blows up the log buffer, /cc @jan-ivar):

- RTCDataChannel.bufferedAmountLowThreshold.html, *multiple*: I'm fine with the whole file being marked as flaky. There aren't many tests in this and they are mostly variants of the same test. If one is flaky, then the attribute is flaky.
- RTCDataChannel-close.html, *Negotiated data channels x10 closing should be synchronized across peers*: Could be grudgingly moved into a new file unless it also fails in Firefox which might indicate that there's a bug in the test.
- RTCDataChannel-send-receive.html, *In-band negotiated channel: Sending and receiving 262144 bytes should succeed or raise TypeError*: This could also be (grudgingly) moved into a different file but I find it misleading. How is it going to be displayed in the wpt.fyi results?
- RTCDataChannel-send-receive.html, *multiple marked as NOTRUN*: This should not happen with the global timeout disabled. I have no idea why it does.

> What exactly does it test? Is it the robustness of the implementation or the compliance of it?

Compliance.

> webrtc/RTCDataChannel-send-receive-stress.html is just one test so I am not sure how feasible splitting is. I would go with just meta/long there.

I can mark it as a *long* test but it will only serve as meta-information and not apply the global timeout of `60 * multiplier` seconds. Perhaps there should be a different meta tag for the purpose of marking it to not run in specific configurations.

foolip commented:
It looks like the Travis job has gotten entirely stuck here, most likely a Travis infrastructure issue since there's not even a link to follow. @lgrahl can you rebase to resolve conflicts? That'll also trigger a new run.

In https://travis-ci.org/web-platform-tests/wpt/builds/444438000, for Chrome Dev the answer is hidden in the log as a Markdown table. This was originally written for a bot to extract and paste as a comment, and here it is:

|                           Test                           |                                                 Subtest                                                 |             Results             |                                                          Messages                                                         |
|----------------------------------------------------------|---------------------------------------------------------------------------------------------------------|---------------------------------|---------------------------------------------------------------------------------------------------------------------------|
| `/webrtc/RTCDataChannel-bufferedAmountLowThreshold.html` | `DOMString of 4 bytes x 10: bufferedamountlow event should fire once bufferedAmount <=\n  2 (custom)`   | **FAIL: 8/10, TIMEOUT: 2/10**   | `assert_unreached: bufferedamountlow event did not fire in time Reached unreachable code;Test timed out`                  |
| `/webrtc/RTCDataChannel-bufferedAmountLowThreshold.html` | `DOMString of 4 bytes x 10: bufferedamountlow event should fire once bufferedAmount <=\n  4 (custom)`   | **NOTRUN: 2/10, TIMEOUT: 8/10** | `Test timed out`                                                                                                          |
| `/webrtc/RTCDataChannel-close.html`                      | `Negotiated data channels x10 closing should be synchronized across peers`                              | **PASS: 6/10, TIMEOUT: 4/10**   | `Test timed out`                                                                                                          |
| `/webrtc/RTCDataChannel-send-receive.html`               | `In-band negotiated channel: Sending and receiving 262144 bytes should succeed or raise\n    TypeError` | **PASS: 2/10, TIMEOUT: 8/10**   | `Test timed out`                                                                                                          |
| `/webrtc/RTCDataChannel-send-receive.html`               | `In-band negotiated channel: Sending and receiving 262144 +1 bytes should raise TypeError`              | **FAIL: 2/10, NOTRUN: 8/10**    | `assert_throws: Should throw TypeError in case the message is too large function "() => dc1.send(message)" did not throw` |
| `/webrtc/RTCDataChannel-send-receive.html`               | `In-band negotiated channel: Sending and receiving 524288 bytes should succeed or raise\n    TypeError` | **NOTRUN: 8/10, TIMEOUT: 2/10** | `Test timed out`                                                                                                          |

Looks like flaky timeouts across the board. Are the tests fairly slow, how long do they take to run locally?

For Firefox, the logs ends like this:
```
Log reached capacity (3145728 bytes); output disabled.

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

This was the topic of https://github.com/web-platform-tests/wpt/issues/11454, which is closed, but clearly it's still a problem.

The stability checks on Taskcluster don't have this problem. @jugglinmike, do you have an ETA on turning off stability checks in Travis?

lgrahl commented:
> Are the tests fairly slow, how long do they take to run locally?

`RTCDataChannel-send-receive.html` takes ~45s for Chrome and `RTCDataChannel-bufferedAmountLowThreshold` ~30s, mainly because of the failures due to timeouts (an event should be fired but it isn't being fired, so we'll need to wait for it for a while).
`RTCDataChannel-close.html` takes only a few seconds for Chrome and the particular test mentioned above was never flaky for me locally. I'm also not seeing anything wrong with it but perhaps someone else wants to take a closer look: https://github.com/web-platform-tests/wpt/blob/99d792a605af57c390a39f7f817dad8135d0c7d2/webrtc/RTCDataChannel-close.html#L152:L184

The `NOTRUN` markings are puzzling me. This should never happen with a local timeout for each test in combination with the global timeout being disabled. Until I know why that happens, I don't think it's useful to run the tests again.

If no one else can look into it, then I will have to put this PR on hold until I have some more spare time.

lgrahl commented:
We can disable the flaky ones but I need to see which ones are flaky for Firefox first. There must be some flaky ones since Travis is failing. @jan-ivar can you or someone else from Mozilla take a look? I am not able to reproduce any *flakyness* on my machine.

> If we start splitting, we could also split modified existing tests/files from new ones.

We can split tests into logical blocks if necessary. I'm not expecting a lot of conflicts and trying to figure out which ones existed prior to this changeset would be a Sisyphean task.

nils-ohlmeier commented:
I found why WPT is logging so much ICE logging: because it uses the default log variables from mozrunner.

I'm currently trying to find a way to change the logging variables in WPT to produce less verbose logging when executed in the WPT environment.

youennf commented:
Looking at https://github.com/web-platform-tests/wpt/issues/11029#issuecomment-389499369 and https://github.com/web-platform-tests/wpt/issues/11120, WPT seems to go away from setting timeout in individual tests.

lgrahl commented:
Note to self: Needs updating to some tests due to https://github.com/w3c/webrtc-pc/pull/1999

alvestrand commented:
@youennf will try to get the impractical (multi-minute) tests disabled or removed; we should merge-and-iterate on the rest.

lgrahl commented:
@alvestrand The main problem seems to be the design choice now (local timeouts for each test, disabled global timeout). Since I have a little bit of time, a suggestion: I'll refactor one of the larger test groups and you folks tell me if it looks good to you. Then I'll proceed with refactoring the rest. Sounds good?

youennf commented:
Current plan is to split this PR as smaller pieces:
- 1 PR for RTCPeerConnection-helper.js changes
- 1 or more PRs for updated test files
- 1 or more PRs for new test files
Some of the new test files may take a very long time.
I am not sure we will be able to land all of them if it cannot be easily run in browsers CI.

lgrahl commented:
To do notes (mostly for myself):

- [ ]  PR for `RTCSctpTransport` changes

jan-ivar commented on webrtc/RTCSctpTransport-onstatechange.html:
```diff
@@ -0,0 +1,56 @@
+ */
+
+promise_test(async (t) => {
+  const resolver = new Resolver();
```
`Resolver` is the deferred anti-pattern, which I'd like to get rid of.

jan-ivar commented on webrtc/RTCSctpTransport-onstatechange.html:
```diff
@@ -0,0 +1,56 @@
+
+  // We use the queue to hold back candidates, so no connection can be established
+  const pc1Queue = new IceCandidateQueue(pc1);
+  const pc2Queue = new IceCandidateQueue(pc2);
```
This being the lone use of *IceCandidateQueue*, I propose using a promise as a simple queue here instead:
```js
promise_test(async (t) => {
  const pc1 = new RTCPeerConnection();
  const pc2 = new RTCPeerConnection();
  t.add_cleanup(() => pc1.close());
  t.add_cleanup(() => pc2.close());
  // We use a queue to hold back candidates, so no connection can be established
  let flush;
  const queue = new Promise(r => flush = r);
  pc1.onicecandidate = e => queue.then(() => pc2.addIceCandidate(e.candidate));
  pc2.onicecandidate = e => queue.then(() => pc1.addIceCandidate(e.candidate));

  // Negotiate SCTP transport
  pc1.createDataChannel('test');
  await doSignalingHandshake(pc1, pc2);

  for (const pc of [pc1, pc2]) {
    assert_equals(pc.sctp.state, 'connecting',
      'RTCSctpTransport should be in the connecting state');
  }
  flush(); // Let the peers connect
  await Promise.all([pc1, pc2].map(async pc => {
    await new Promise(r => pc.sctp.onstatechange = r);
    assert_equals(pc.sctp.state, 'connected');
  }));
}
```
This should also fix that the test doesn't bail before both `pc`s have had their `statechange` event tested.

jan-ivar replied:
@lgrahl Odd, looks like github had some old review comments cached here. The last comment is the one I wanted you to look at.

jan-ivar replied:
@youennf ^

jan-ivar replied:
The first unresolved comment thread is the one about splitting this up. Presumably that's still happening? I just commented here because I couldn't find this test in the newer PRs. If we split it up and avoid modifying *exchangeIceCandidates* to take a queue, then I'm happy.

alvestrand commented:
Status Dec 18: Waiting for split-up PRs to land, then we'll close this one.

alvestrand commented:
Closing this PR on the assumption that split-up PRs have landed.
